servers/rendering/rendering_device.cpp

```diff
@@ -91,771 +442,6324 @@ RID RenderingDevice::shader_create_from_spirv(const Vector<ShaderStageSPIRVData>
+	ERR_FAIL_COND_V(!index_buffer.driver_id, RID());
+
+	if (p_data.size()) {
+		uint64_t data_size = p_data.size();
+		const uint8_t *r = p_data.ptr();
+		_buffer_update(&index_buffer, 0, r, data_size);
+		RDD::BufferBarrier bb;
+		bb.buffer = index_buffer.driver_id;
+		bb.src_access = RDD::BARRIER_ACCESS_TRANSFER_WRITE_BIT;
+		bb.dst_access = RDD::BARRIER_ACCESS_INDEX_READ_BIT;
+		bb.size = data_size;
+		driver->command_pipeline_barrier(frames[frame].setup_command_buffer, RDD::PIPELINE_STAGE_TRANSFER_BIT, RDD::PIPELINE_STAGE_VERTEX_INPUT_BIT, {}, bb, {});
+	}
+
+	buffer_memory += index_buffer.size;
+
+	RID id = index_buffer_owner.make_rid(index_buffer);
+#ifdef DEV_ENABLED
+	set_resource_name(id, "RID:" + itos(id.get_id()));
+#endif
+	return id;
+}
+
+RID RenderingDevice::index_array_create(RID p_index_buffer, uint32_t p_index_offset, uint32_t p_index_count) {
+	_THREAD_SAFE_METHOD_
+
+	ERR_FAIL_COND_V(!index_buffer_owner.owns(p_index_buffer), RID());
+
+	IndexBuffer *index_buffer = index_buffer_owner.get_or_null(p_index_buffer);
+
+	ERR_FAIL_COND_V(p_index_count == 0, RID());
+	ERR_FAIL_COND_V(p_index_offset + p_index_count > index_buffer->index_count, RID());
+
+	IndexArray index_array;
+	index_array.max_index = index_buffer->max_index;
+	index_array.driver_id = index_buffer->driver_id;
+	index_array.offset = p_index_offset;
+	index_array.indices = p_index_count;
+	index_array.format = index_buffer->format;
+	index_array.supports_restart_indices = index_buffer->supports_restart_indices;
+
+	RID id = index_array_owner.make_rid(index_array);
+	_add_dependency(id, p_index_buffer);
+	return id;
+}
+
+/****************/
+/**** SHADER ****/
+/****************/
+
+static const char *SHADER_UNIFORM_NAMES[RenderingDevice::UNIFORM_TYPE_MAX] = {
+	"Sampler", "CombinedSampler", "Texture", "Image", "TextureBuffer", "SamplerTextureBuffer", "ImageBuffer", "UniformBuffer", "StorageBuffer", "InputAttachment"
+};
+
+String RenderingDevice::_shader_uniform_debug(RID p_shader, int p_set) {
+	String ret;
+	const Shader *shader = shader_owner.get_or_null(p_shader);
+	ERR_FAIL_NULL_V(shader, String());
+	for (int i = 0; i < shader->uniform_sets.size(); i++) {
+		if (p_set >= 0 && i != p_set) {
+			continue;
+		}
+		for (int j = 0; j < shader->uniform_sets[i].size(); j++) {
+			const ShaderUniform &ui = shader->uniform_sets[i][j];
+			if (!ret.is_empty()) {
+				ret += "\n";
+			}
+			ret += "Set: " + itos(i) + " Binding: " + itos(ui.binding) + " Type: " + SHADER_UNIFORM_NAMES[ui.type] + " Writable: " + (ui.writable ? "Y" : "N") + " Length: " + itos(ui.length);
+		}
+	}
+	return ret;
+}
+
+String RenderingDevice::shader_get_binary_cache_key() const {
+	return driver->shader_get_binary_cache_key();
+}
+
+Vector<uint8_t> RenderingDevice::shader_compile_binary_from_spirv(const Vector<ShaderStageSPIRVData> &p_spirv, const String &p_shader_name) {
+	ShaderReflection reflection;
+	if (_reflect_spirv(p_spirv, reflection) != OK) {
+		return Vector<uint8_t>();
+	}
```

Comment:
I need to use SPIRV-Cross for Metal, in order to generate MSL from the SPIR-V. I realise the `_reflect_spirv` does some validation, but otherwise, I don't really need it, so it is wasted CPU.

Does it make sense to leave it to the driver to decide if it wants to use `_reflect_spirv`?

Reply:
Very good point. In the next push, I'm making `_reflect_spirv()` an optional service that drivers can choose to use or not.

Reply:
Disregard, as I see I'll use this as a sanity chec